src/mlpack/methods/kde/kde_rules_impl.hpp

```diff
@@ -125,22 +132,38 @@ Score(const size_t queryIndex, TreeType& referenceNode)
     const math::Range r = referenceNode.RangeDistance(queryPoint);
     minDistance = r.Lo();
     maxDistance = r.Hi();
+
+    // Check if we are a self-child.
+    if (tree::TreeTraits<TreeType>::HasSelfChildren &&
+        referenceNode.Parent() != NULL &&
+        referenceNode.Parent()->Point(0) == referenceNode.Point(0))
+    {
+      alreadyDidRefPoint0 = true;
+    }
```

Comment:
In this case, technically we have already done the base case between the query point and `referenceNode.Point(0)`, but I couldn't think of an easy way to cache that and pass it down the tree.  So it's unfortunate but it seems there is some duplicated computation here.  (I don't think it's worth solving in this PR, but I at least wanted to point it out.)